crates/ide-assists/src/handlers/convert_closure_to_fn.rs

```diff
@@ -0,0 +1,1258 @@
+        {
+            wrap_body_in_block = false;
+        }
+    };
+
+    acc.add(
+        AssistId("convert_closure_to_fn", AssistKind::RefactorRewrite),
+        "Convert closure to fn",
+        text_range,
+        |builder| {
+            let closure_name_or_default = closure_name
+                .as_ref()
+                .map(|(_, _, it)| it.clone())
+                .unwrap_or_else(|| make::name("fun_name"));
+            let captures = closure_ty.captured_items(ctx.db());
+            let capture_tys = closure_ty.capture_types(ctx.db());
+
+            let mut captures_as_args = Vec::with_capacity(captures.len());
+
+            let body_root = body.syntax().ancestors().last().unwrap();
+            // We need to defer this work because otherwise the text range of elements is being messed up, and
+            // replacements for the next captures won't work.
+            let mut capture_usages_replacement_map = Vec::with_capacity(captures.len());
+
+            for (capture, capture_ty) in std::iter::zip(&captures, &capture_tys) {
+                // FIXME: Allow configuring the replacement of `self`.
+                let capture_name =
+                    if capture.local().is_self(ctx.db()) && !capture.has_field_projections() {
+                        make::name("this")
+                    } else {
+                        make::name(&capture.place_to_name(ctx.db()))
+                    };
+
+                closure_mentioned_generic_params.extend(capture_ty.generic_params(ctx.db()));
+
+                let capture_ty = capture_ty
+                    .display_source_code(ctx.db(), module.into(), true)
+                    .unwrap_or_else(|_| "_".to_owned());
+                params.push(make::param(
+                    ast::Pat::IdentPat(make::ident_pat(false, false, capture_name.clone_subtree())),
+                    make::ty(&capture_ty),
+                ));
+
+                for capture_usage in capture.usages().sources(ctx.db()) {
+                    if capture_usage.file_id() != ctx.file_id() {
+                        // This is from a macro, don't change it.
+                        continue;
+                    }
+
+                    let capture_usage_source = capture_usage.source();
+                    let capture_usage_source = capture_usage_source.to_node(&body_root);
+                    let expr = match capture_usage_source {
+                        Either::Left(expr) => expr,
+                        Either::Right(pat) => {
+                            let Some(expr) = expr_of_pat(pat) else { continue };
+                            expr
+                        }
+                    };
+                    let replacement = wrap_capture_in_deref_if_needed(
+                        &expr,
+                        &capture_name,
+                        capture.kind(),
+                        capture_usage.is_ref(),
+                    )
+                    .clone_for_update();
```

Comment:
This is more a question for @Veykril, but: is there an alternative to the `clone_for_update` call that could be used here or do we need to wait to wait for https://github.com/rust-lang/rust-analyzer/issues/15710#issuecomment-2294892785 to happen first?

Reply:
There are a couple, but they are all likewise kind of annoying to do ... I imagine until we get the node tracking stuff making this not use mutable nodes is gonna be ugly